interfaces/builtin/cups_control.go

```diff
@@ -19,30 +19,159 @@
 
 package builtin
 
+import (
+	"strings"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/release"
+	"github.com/snapcore/snapd/snap"
+)
+
+// On classic systems where the slot is implicitly provided, the interface
+// allows access to the cups control socket.
+//
+// On systems where the slot is provided by an app snap, the cups-control
+// interface is the companion interface to the cups interface. The design of
+// these interfaces is based on the idea that the slot implementation (eg
+// cupsd) is expected to query snapd on if the cups-control interface is
+// connected or not and the print service will mediate admin functionality (ie,
+// the rules in these interfaces allow connecting to the print service, but do
+// not implement enforcement rules; it is up to the print service to provide
+// enforcement).
 const cupsControlSummary = `allows access to the CUPS control socket`
 
+// cups-control is implicit on classic but may also be provided by an app snap
+// on core or classic (the current design allows the snap provider to slots
+// both cups-control and cups or just cups-control (like with implicit classic
+// or any slot provider without mediation patches), but not just cups).
 const cupsControlBaseDeclarationSlots = `
   cups-control:
     allow-installation:
       slot-snap-type:
+        - app
         - core
     deny-auto-connection: true
+    deny-connection:
+      on-classic: false
```

Comment:
so to clarify for everybody, this interface was also before implicit on classic, it wasn't and isn't implicit on core. Now app-provided slots can exist for it.

Before manual connection was allowed on classic and core but on core no slot could actually exist.
Now slots can exist on both core and classic,  on classic provided by system or app, on core provided by app.

This allows manual connection on classic to either I think, but stops it on core...

* it needs a test I think in base_declaration_test.go
* another possibility would be to prohibit in both places manual connection if the slot is from an app?  (we don't seem ATM to allow the syntax for that but we could... (we support only plug-snap-types in slot connection constraints, but we could support slot-snap-types as well in theory)

Reply:
Note, for the base declaration changes to cups-control, I followed the pattern we used for avahi-observe/avahi-control since it is the same: it can be implicit on classic or app installed on core and classic. It cannot be implicit on core. (This is in contrast to the new cups interface which today can only be via app snap).

@anonymouse64, this is why we made the distinction here but you are right that it is somewhat awkward. What we are trying to achieve is not requiring a connection constraint via snap declaration for an implicit slot, and this does that. It does a little more though since it grants connection to an app snap on classic. This is 'ok' because the review-tools understand this established pattern and will flag for manual review a snap that is trying to slots cups-control. We might want to do something like this:

```
const cupsControlBaseDeclarationSlots = ` 
  cups-control:
    allow-installation:
      slot-snap-type:
        - app
        - core
    deny-auto-connection: true
    allow-connection:
      on-classic: true
      slot-snap-type:
      - core
```

but as @pedronis mentioned, `slot-snap-type` is not allowed in a connection constraint for `slots`. `slot-snap-type` is allowed in the connection constraint for `plugs` though, so we could use this in the base declaration:

```
const cupsControlBaseDeclarationSlots = ` 
  cups-control:
    allow-installation:
      slot-snap-type:
        - app
        - core

const cupsControlBaseDeclarationPlugs = `
  cups-control:
    allow-installation:
      plug-snap-type:
        - app
    deny-connection:
      on-classic: true
      slot-snap-type:
        - app
    deny-auto-connection: true
```

which is more accurate, but it falls outside of our established patterns and is arguably more difficult to understand (note, the review-tools would still flag something slotting cups-control with the above for manual review). IME, if we determine that we want a new pattern for things like avahi and cups, then those base declaration changes should be in a followup PR. I think I would prefer adding 'slot-snap-types' to the connection constraint instead of contorting the existing syntax like this though.

@pedronis - the tests were updated in the same manner as for avahi-control/avahi-observe by adding `cups-control` to `slotInstallation` in basedeclaration_test.go. I'll take a look at adding more tests for this pattern (avahi* and cups-control).

Reply:
I unresolved this convo because otherwise the link added in the tests comment doesn't get here